server/src/main/java/io/druid/server/ClientInfoResource.java

```diff
@@ -83,13 +100,83 @@ public ClientInfoResource(
   @Produces(MediaType.APPLICATION_JSON)
   public Map<String, Object> getDatasource(
```

Comment:
Pragmatically speaking, when it comes to http codes, most popular ones are..
200, 400, 404 and 500
and they generally cover most cases and well understood by everyone. 204 with empty response will be treated exactly same as 200 with empty response. there is hardly any difference.
If we want to be strict, we should do 400 or else we should just do 200 with empty response.

Personally, I prefer 200 with empty response because it may happen that a datasource exists but none of its segments are really loaded in the cluster and broker might not know about it. So semantically,  saying that "nothing from this datasource is being served" is OK.